src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
-                case TokenKind.QuestionQuestionEquals: et = ExpressionType.Coalesce; break;
+                return assignmentTarget.SetValue(this, right);
             }
 
+            var expressionType = tokenKind switch
+            {
+                TokenKind.PlusEquals => ExpressionType.Add,
+                TokenKind.MinusEquals => ExpressionType.Subtract,
+                TokenKind.MultiplyEquals => ExpressionType.Multiply,
+                TokenKind.DivideEquals => ExpressionType.Divide,
+                TokenKind.RemainderEquals => ExpressionType.Modulo,
+                TokenKind.QuestionQuestionEquals => ExpressionType.Coalesce,
+                _ => ExpressionType.Extension
+            };
+
+            return ReduceCompoundAssignment(this, assignmentTarget, lhsMemberAccessAst, expressionType, right);
+        }
+
+        internal static BlockExpression ReduceCompoundAssignment(
+            Compiler compiler,
+            IAssignableValue assignmentTarget,
+            MemberExpressionAst lhsMemberExpressionAst,
+            ExpressionType expressionType,
+            Expression valueToAssign)
+        {
             var exprs = new List<Expression>();
             var temps = new List<ParameterExpression>();
-            var getExpr = av.GetValue(this, exprs, temps);
+            var getExpr = assignmentTarget.GetValue(compiler, exprs, temps);
 
-            if (et == ExpressionType.Coalesce)
+            if (expressionType == ExpressionType.Coalesce)
             {
-                exprs.Add(av.SetValue(this, Coalesce(getExpr, right)));
+                exprs.Add(assignmentTarget.SetValue(compiler, Coalesce(getExpr, valueToAssign)));
+            }
+            else if (lhsMemberExpressionAst != null
+                && (expressionType == ExpressionType.Add || expressionType == ExpressionType.Subtract))
+            {
+                // We're essentially compiling += / -= expressions as a conditional here if the LHS is a member access
+                // expression.
+                //
+                // The expression `$a.Member += $b` will be compiled more like this:
+                //
+                //      $tmp = $a.Member
+                //      if ($tmp -is [PSEvent]) {
+                //          $tmp.Value.GetAddMethod().Invoke($b)
+                //      }
+                //      else {
+                //          $a.Member = $tmp + $b
+                //      }
+                //
+                // The equivalent code is very similar for -= with the exception of calling GetRemoveMethod() instead of
+                // GetAddMethod().This lets us handle event delegates with += or -=, but we should only make this check
+                // for the += / -= compound operators.
+                ParameterExpression lhsStoreVar = Expression.Variable(typeof(object));
+                temps.Add(lhsStoreVar);
+
+                exprs.Add(Expression.Assign(lhsStoreVar, getExpr.Cast(typeof(object))));
```

Comment:
What is the cast needed for here?

Reply:
I don't recall... I'll try it without, but I probably copied this pattern from elsewhere in the compiler.

Reply:
@daxian-dbw may know why it's needed

Reply:
> What is the cast needed for here?

More than likely it's because expressions don't implicitly box, the cast here will do that.  Also matching expected types is super finicky (like a `Expression.Block`'s return type, or a variable) which casting everything to `object` greatly simplifies.

That's why there's a ton of casting, but I don't know off the top of my head of `getExpr` would already be cast. (Side note, it's also why byref/likes will probably never be supported 🙁)